### PR TITLE
checker: merge-region should consider the peer on an offline store

### LINKED ISSUE #4119: merge should consider offline store
### What did you do?

Delete a store by
```
store delete store_id
```

### What did you expect to see?

No new replicas are scheduled to the deleted stores.

### What did you see instead?

New replicas are schedule to deleted stores from time to time. Apparently, merge scheduler doesn't if stores are offline or not.
```
[2021/09/14 07:50:14.593 +00:00] [INFO] [operator_controller.go:543] ["operator finish"] [region-id=208025] [takes=6.063414281s] [operator="\"replace-rule-offline-leader-peer {mv peer: store [367567] to [6]} (kind:replica,region,leader, region:208025(2144,851), createAt:2021-09-14 07:50:08.530475804 +0000 UTC m=+324583.120329687, startAt:2021-09-14 07:50:08.530563423 +0000 UTC m=+324583.120417305, currentStep:5, steps:[add learner peer 532133 on store 6, transfer leader from store 367567 to store 5, use joint consensus, promote learner peer 532133 on store 6 to voter, demote voter peer 480732 on store 367567 to learner, leave joint state, promote learner peer 532133 on store 6 to voter, demote voter peer 480732 on store 367567 to learner, remove peer on store 367567]) finished\""] [additional-info=]
[2021/09/14 07:50:50.948 +00:00] [INFO] [operator_controller.go:430] ["add operator"] [region-id=208025] [operator="\"merge-region {merge: region 208025 to 516311} (kind:merge,region, region:208025(2144,857), createAt:2021-09-14 07:50:50.94888579 +0000 UTC m=+324625.538739654, startAt:0001-01-01 00:00:00 +0000 UTC, currentStep:0, steps:[add learner peer 532180 on store 367567, use joint consensus, promote learner peer 532180 on store 367567 to voter, demote voter peer 532133 on store 6 to learner, leave joint state, promote learner peer 532180 on store 367567 to voter, demote voter peer 532133 on store 6 to learner, remove peer on store 6, merge region 208025 into region 516311])\""] [additional-info=]
```

### What version of PD are you using (`pd-server -V`)?
```
$ ./deploy/pd-2379/bin/pd-server --version
Release Version: v5.1.1
Edition: Community
Git Commit Hash: 7cba1912b317a533e18b16ea2ba9a14ed2891129
Git Branch: heads/refs/tags/v5.1.1
UTC Build Time:  2021-07-23 01:16:18
```

### PR DESCRIPTION
**Pull request by JmPotato**
Signed-off-by: JmPotato <ghzpotato@gmail.com>

### What problem does this PR solve?

Close #4119.

### What is changed and how it works?

Make `merge_checker.go` check whether there is a peer of the adjacent region on an offline store, while the source region has no peer on it. This is to prevent from bringing any other peer into an offline store to slow down the offline process.

### Check List

Tests

- Unit test
- Integration test

### Release note

```release-note
Make merge-region not to merge a region into the offline store.
```

### CODE REVIEW

**Comment by rleungx** [comment]
I think it doesn't solve the original issue since in that issue the store is still existed but in the offline state.

**Comment by JmPotato** [comment]
> I think it doesn't solve the original issue since in that issue the store is still existed but in the offline state.

I see...

**Comment by JmPotato** [comment]
> I think it doesn't solve the original issue since in that issue the store is still existed but in the offline state.

Fixed. PTAL, thx!

**Review by rleungx — APPROVED** [review]

File: server/schedule/checker/merge_checker.go [hunk 1]
```diff
@@ -145,6 +145,13 @@ func (m *MergeChecker) Check(region *core.RegionInfo) []*operator.Operator {
 		return nil
 	}
 
+	// Check whether the store of the target region leader is normal.
+	store := m.cluster.GetStore(target.GetLeader().GetStoreId())
```
Comment (disksing) [inline server/schedule/checker/merge_checker.go:149]: why only consider leader instead of all peers?

Comment (JmPotato) [reply server/schedule/checker/merge_checker.go:149]: After merging, the target Region will extend its range and provide the service, so I think we just need to make sure the leader of the target Region is in a normal store?

Comment (disksing) [reply server/schedule/checker/merge_checker.go:149]: But the original issue did not mention anything about the leader. It looks the main issue is moving peer to offline store may slow down the offline process.

File: server/schedule/checker/merge_checker.go [hunk 2]
```diff
@@ -222,6 +222,21 @@ func AllowMerge(cluster opt.Cluster, region *core.RegionInfo, adjacent *core.Reg
...
+
+// Check whether there is a peer of the adjacent region on an offline store,
+// while the source region has no peer on it. This is to prevent from bringing
+// any other peer into an offline store to slow down the offline process.
+func checkPeerStore(cluster opt.Cluster, region, adjacent *core.RegionInfo) bool {
+	regionStoreIDs := region.GetStoreIds()
+	for _, peer := range adjacent.GetPeers() {
+		storeID := peer.GetStoreId()
+		store := cluster.GetStore(storeID)
+		if _, ok := regionStoreIDs[storeID]; (store == nil || store.IsOffline()) && !ok {
```
Comment (disksing) [inline server/schedule/checker/merge_checker.go:237]:
```
if store == nil || store.IsOffline {
    if _, ok := ...; !ok {
    ...
}
```

Comment (disksing) [reply server/schedule/checker/merge_checker.go:237]: Other parts look file!

**Review by disksing — APPROVED** [review]

**Pull request merged**